Establish a docs framework - Resolves #60 (#84)
grafana/dashboard-linter: docs/index.md

* `panel-datasource-rule` - Checks that each panel uses the templated datasource.
* `panel-title-description-rule` - Checks that each panel has a title and description.
* `panel-units-rule` - Checks that each panel uses has valid units defined.
* `target-promql-rule` - Checks that each target uses a valid PromQL query.
* `target-rate-interval-rule` - Checks that each target uses $__rate_interval.
* `target-job-rule` - Checks that every PromQL query has a job matcher.
* `target-instance-rule` - Checks that every PromQL query has a instance matcher.

## Related Rules

There are groups of rules that are intended to drive certain outcomes, but may be implemented separately to allow more granular [exceptions](#exclusions-and-warnings), and to keep the rules terse.

## Job and Instance Template Variables

The following rules work together to ensure that every dashboard has template variables for Job and Instance, and that they are properly configured, and used in every promql query.

* `template-job-rule`
* `template-instance-rule`
* `target-job-rule`
* `target-instance-rule`

The reasoning is that.. WIP, but...
* All prom metrics will have these, as it's part of the spec

> * Other reasons?

Author: @v-zhuravlev can you help provide some context here? I know you've raised concerns about enforcing these rules, so if we can have that discussion in the comments here, or flesh-out this document it would be appreciated.

Reviewer: As it turns out, there is a problem with $job (and $instance as well) filter becomes visible when you start to compose dashboards that could have panels from multiple sources. For example, prometheus **node_exporter** and promtail logs from the same host. Or from **node_exporter** and **process_exporter** exporters that reside on the same host.

I think that at first the idea of this rule was to have a two-level grouping of instances:
- 1st level: instance
- 2nd level: job

This is convenient, but since job is [built-in label](https://prometheus.io/docs/concepts/jobs_instances/#jobs-and-instances), using it for grouping/filtering makes prometheus/grafana-agent configs complex (if you have multiple metrics/logs sources):
- It requires rewriting already present labels, so the match for metrics&logs , making scrape_configs complex due to additional relabelling rules. 
- Since the label is present, relabelling seems to be the only way to overwrite it. If this label were vacant then we could just used _external_labels_ or _labels_ to set the common job in grafana-agent/prometheus.

Here are some examples of configs where such user confusing relabeling taking place:
https://grafana.com/docs/grafana-cloud/integrations/integrations/integration-docker/#post-install-configuration-for-the-docker-integration
https://grafana.com/docs/grafana-cloud/integrations/integrations/integration-mysql/

So, we should ask, what is the main reason we enforce this?
If we need a second level of grouping, perhaps we should use label like `cluster` instead? (and make it optional, where it is applicable)

Author: I went ahead and documented this in it's "current state" to get this document framework committed.

Let's revisit the effectiveness of these rules, and we can address in a follow-up PR. I'll create an issue to track.

  lint        Lint a dashboard
  rules       Print documentation about each lint rule.

Flags:
  -h, --help   help for dashboard-linter

Use "dashboard-linter [command] --help" for more information about a command.
```

## Completion
[embedmd]:# (_intermediate/completion.txt)
```txt
Generate the autocompletion script for dashboard-linter for the specified shell.
See each sub-command's help for details on how to use the generated script.

Usage:
  dashboard-linter completion [command]

Available Commands:
  bash        Generate the autocompletion script for bash
  fish        Generate the autocompletion script for fish
  powershell  Generate the autocompletion script for powershell
  zsh         Generate the autocompletion script for zsh

Flags:
  -h, --help   help for completion

Use "dashboard-linter completion [command] --help" for more information about a command.
```

## Lint
[embedmd]:# (_intermediate/lint.txt)
```txt
Returns warnings or errors for dashboard which do not adhere to accepted standards

Usage:
  dashboard-linter lint [dashboard.json] [flags]

Flags:
  -h, --help      help for lint
      --strict    fail upon linting error or warning

>       --verbose   show more information about linting

Reviewer 2: Does verbose show the readme definitions for the rules added here or do we have to refer to github for that?

Author: No, verbose just shows *every* result, including success. By default, success messages are suppressed. This option is like the opposite of `quiet` which does not exist as an option, since it's the default behavior.

One current example is with the [Grafana Cloud Docker Integration](https://grafana.com/docs/grafana-cloud/data-configuration/integrations/integration-reference/integration-docker/#post-install-configuration-for-the-docker-integration), which combines metrics from cAdvisor, and logs from the docker daemon using `docker_sd_configs`.

In this case, without label rewriting the logs would not have any labels at all. The relabeling on the metrics is in order to apply an opinionated job name rather than the default the agent would provide (`integrations/cadvisor`).

For dashboards like this, it is advised that a linting [exception](#exclusions-and-warnings) be created for these rules, and a separate label which exists on data from all data sources be used to filter.

# Exclusions and Warnings

Where the rules above don't make sense, you can drop a `.lint` file in a same directory as the dashboard telling the linter to ignore certain rules, or downgrade them to warnings.

Example:
```yaml
exclusions:
  template-job-rule:
warnings:
  template-instance-rule:
```

## Reasons

> It is advised that whenever you exclude, or warn for a rule, that you provide a reason. This allows for other maintainers of your dashboard to understand why a particular rule may not be followed. Eventually the dashboard-linter will provide reporting that echos that reason back to the user.

Reviewer 3: ```suggestion
Whenever you exclude or warn for a rule, it's recommended that you provide a reason. This allows for other maintainers of your dashboard to understand why a particular rule may not be followed. Eventually, the dashboard-linter will provide reporting that echoes that reason back to the user.
```
